### PR TITLE
Added conflict detection to Shortcut Mapper

### PR DESCRIPTION
**Pull request by A-R-C-A**
Added a basic conflict detection functionality to the Shortcut Mapper. This includes:
* Marking conflicts with a different background color.
* Displaying conflict information.
* Warning when attempting to assign an already existing short-key.

Some more related changes to the Shortcut Mapper:
* Fixed some bugs, for instance:
  * Babygrid was shrinking on any attempt to adjust its position to integral rows.
  * Context menu entries were enabled for empty lists. Using them would crash Npp.
  * ... some more minor bug fixes.
* Performance enhancements to Babygrid.
* Minor UI improvements.
* Babygrid is configured dpi aware.

### CODE REVIEW

File: PowerEditor/src/WinControls/Grid/BabyGrid.cpp [hunk 1]
```diff
@@ -462,11 +468,14 @@ void DisplayColumn(HWND hWnd,int SI,int c,int offset,HFONT hfont,HFONT hcolumnhe
 		 rectsave=rect;
 		 SetCell(&BGcell,r,c);
 		 lstrcpy(buffer, TEXT(""));
-		 SendMessage(hWnd,BGM_GETCELLDATA,(WPARAM)&BGcell,(LPARAM)buffer);
 		 if((c==0)&&(BGHS[SI].ROWSNUMBERED))
 		 {
 		  wsprintf(buffer, TEXT("%d"), r);
+		  iProperty = 2 << 4;
```
Comment (MAPJe71) [inline PowerEditor/src/WinControls/Grid/BabyGrid.cpp]: I would like to see some comment here to explain the 'magic' numbers.

File: PowerEditor/src/WinControls/Grid/BabyGrid.cpp [hunk 2]
```diff
@@ -1741,11 +1790,26 @@ LRESULT CALLBACK GridProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
...
+					}
+				   switch (buffer[11])
+				    {
+				     case 'A': ReturnValue |= 1 << 4; break;
+				     case 'N': ReturnValue |= 2 << 4; break;
+				     case 'T': ReturnValue |= 3 << 4; break;
+				     case 'F': ReturnValue |= 4 << 4; break;
+				     case 'G': ReturnValue |= 5 << 4; break;
+				     default : ReturnValue |= 1 << 4; break;
+				    }
```
Comment (MAPJe71) [inline PowerEditor/src/WinControls/Grid/BabyGrid.cpp]: Again, please add comment to explain the 'magic' numbers.

File: PowerEditor/src/WinControls/Grid/ShortcutMapper.h [hunk 3]
```diff
@@ -91,9 +94,33 @@ protected :
 
 	TCHAR tabNames[5][maxTabName];
 
+	//save/restore the last view
+	std::vector<size_t> _lastHomeRow { 1, 1, 1, 1, 1 };
+	std::vector<size_t> _lastCursorRow { 1, 1, 1, 1, 1 };
+
+	const generic_string _defaultInfo = TEXT("No schortcut conflicts for this item.");
```
Comment (MAPJe71) [inline PowerEditor/src/WinControls/Grid/ShortcutMapper.h]: I believe it's spelled "shortcut".

**Comment by A-R-C-A** [comment]
I'm not the author of this 'magic' numbers. Buffer[10] was copied from BGM_GETPROTECTION and buffer[11] was copied from BGM_GETTYPE. This changes here will reduce 3 message calls per cell (with a 'heavy' task like BinarySearchListBox(...)) to only 1 call per cell, while painting/refreshing the view. This will lead to a better performance. Maybe, I'll add some comments to explain my intention a bit more.

**Comment by MAPJe71** [comment]
Aha thanx, now I get it ... you combined the return values of BGM_GETPROTECTION and BGM_GETTYPE and added it as return value for BGM_GETCELLDATA.

FYI:

    buffer[0] .. buffer[4]		row number
    buffer[5]					hyphen as field separator
    buffer[6] .. buffer[9]		column number
    buffer[9]					pipe character as field separator
    buffer[10]					protection indicator either 'P' (protected) or 'U' (unprotected)
    buffer[11]					data type indicator one of 'A' (alpha i.e. text), 'N' (numeric), 'T' (boolean true), 'F' (boolean false) and 'G' (graphic)
    buffer[12]					pipe character as field separator
    buffer[13] .. buffer[n]		cell data

**Comment by A-R-C-A** [comment]
Yep

**Comment by donho** [comment]
https://github.com/notepad-plus-plus/notepad-plus-plus/commit/1abac157993685c4f6345443f6256403526858ad

**Pull request closed without merging**